### PR TITLE
API-key-rotation-update-api-ke

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

**Review by ambiknai — COMMENTED** [review]
lgtm

File: go.mod [hunk 1]
```diff
@@ -4,11 +4,12 @@ go 1.15
 
 require (
 	github.com/IBM-Cloud/ibm-cloud-cli-sdk v0.6.7
-	github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8
+	github.com/IBM/ibmcloud-volume-interface v1.0.1-0.20211130042446-32c5e8b6904f
```
Comment (sameshai) [inline go.mod:7]: Once we push the interface PR please make sure we update the release

**Review by sameshai — APPROVED** [review]
LGTM

**Comment by contsto** [comment]
:red_circle: Coverage decreased from [90.3279%] to [90.0382%]

**Comment by contsto** [comment]
:red_circle: Coverage decreased from [90.3279%] to [89.9676%]

**Comment by contsto** [comment]
:red_circle: Coverage decreased from [90.3279%] to [89.9676%]

File: go.mod [hunk 2]
```diff
@@ -4,11 +4,18 @@ go 1.15
...
-	github.com/stretchr/testify v1.6.1
+	github.com/stretchr/testify v1.7.0
 	go.uber.org/zap v1.15.0
 	golang.org/x/net v0.0.0-20200707034311-ab3426394381
 )
+
+replace (
+	k8s.io/apiserver => k8s.io/apiserver v0.0.0-20190313205120-8b27c41bdbb1
+	k8s.io/component-base => k8s.io/component-base v0.0.0-20190617074208-2b0aae80ca81
+)
```
Comment (prankulmahajan) [inline go.mod:21]: I think we can remove this, since we have removed csi-common reference

**Comment by contsto** [comment]
:red_circle: Coverage decreased from [90.3279%] to [89.9676%]

**Comment by contsto** [comment]
:red_circle: Coverage decreased from [90.3279%] to [89.9676%]

**Comment by contsto** [comment]
:thumbsup: Coverage increased from [90.3279%] to [90.4235%]

**Comment by contsto** [comment]
:thumbsup: Coverage increased from [90.3279%] to [90.3926%]

**Comment by GunaKKIBM** [comment]
Test results:

Creating pvc by deleting api key, initially, the pvc is in pending state, kubernetes retries to create volume. Meanwhile, if the api key is reset, volume gets created. It won't be necessary to restart the pod. 
```
kubectl create -f pvc.yaml                                   
persistentvolumeclaim/csi-vpc-block-5iops-pvc-api-key-rotation40 created

gunakk@Gunas-MBP OPENSHIFT_VPC_CLUSTER_TESTING % kubectl describe pvc csi-vpc-block-5iops-pvc-api-key-rotation40 
Name:          csi-vpc-block-5iops-pvc-api-key-rotation40
Namespace:     default
StorageClass:  ibmc-vpc-block-5iops-tier
Status:        Bound
Volume:        pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                 Age                   From                                                                                      Message
  ----     ------                 ----                  ----                                                                                      -------
  Warning  ProvisioningFailed     3m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 9d2c7752-c1e4-491a-a61f-fbb45a30b7d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Warning  ProvisioningFailed     2m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 077e111a-e15c-46d4-a54e-0abc0cf919d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   Provisioning           88s (x4 over 4m29s)   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  External provisioner is provisioning volume for claim "default/csi-vpc-block-5iops-pvc-api-key-rotation40"
  Warning  ProvisioningFailed     88s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: b1c7a84a-04ec-4f74-be16-47ef86202d18 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   ExternalProvisioning   20s (x18 over 4m29s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal   ProvisioningSucceeded  11s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  Successfully provisioned volume pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c

 % kubectl create -f validPOD2.yaml
deployment.apps/api-key-rotation-pod40 created

 % kubectl get pods
NAME                                      READY   STATUS             RESTARTS   AGE
api-key-rotation-pod40-6fff7bb57-vmnxn    1/1     Running            0          3m26s
```

File: block/utils/init_provider.go [hunk 3]
```diff
@@ -70,35 +69,32 @@ func InitProviders(conf *vpcconfig.VPCBlockConfig, logger *zap.Logger) (registry
...
 	}
 	ctxLogger.Info("Calling provider/utils/init_provider.go GenerateContextCredentials")
-	contextCredentials, err := GenerateContextCredentials(conf, providerID, ccf, ctxLogger)
+	contextCredentials, err := GenerateContextCredentials(vpcBlockConfig, providerID, ccf, ctxLogger)
+	if err != nil {
+		ctxLogger.Error("Unable to generate credentials", local.ZapError(err))
+		return nil, true, err
+	}
+	session, err := prov.OpenSession(ctx, contextCredentials, ctxLogger)
 	if err == nil {
```
Comment (arahamad) [inline block/utils/init_provider.go:91]: make it uniform accros other conditions

File: go.sum [hunk 4]
```diff
@@ -2,8 +2,8 @@ github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/IBM-Cloud/ibm-cloud-cli-sdk v0.6.7 h1:eHgfQl6IeSmzWUyiSi13CvoFYsovoyqWlpHX0pa9J54=
 github.com/IBM-Cloud/ibm-cloud-cli-sdk v0.6.7/go.mod h1:RiUvKuHKTBmBApDMUQzBL14pQUGKcx/IioKQPIcRQjs=
-github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8 h1:0oCBQidejAlLI5f5PlBHIytArhJI1aEOBCCOMoxNOPs=
-github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8/go.mod h1:2LxqkERml3eBxfRkQniuiRmKkAnl5tfC0ttlP2gupJM=
+github.com/IBM/ibmcloud-volume-interface v1.0.1-0.20211203045519-74fe99e68e4f h1:XcW0pted1N9FWNhtspx38USjCmWduICvngyeU/OZ34g=
```
Comment (arahamad) [inline go.sum:5]: update this

**Review by arahamad — CHANGES_REQUESTED** [review]
pl check comments

**Review by prankulmahajan — APPROVED** [review]
lgtm

**Comment by contsto** [comment]
:thumbsup: Coverage increased from [90.3279%] to [90.3926%]

**Review by arahamad — APPROVED** [review]
lgtm

**Pull request merged**